drivers/net/native_sim.h

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright (c) 2023 Marcin Niestroj
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __DRIVERS_NET_NATIVE_SIM_H__
+#define __DRIVERS_NET_NATIVE_SIM_H__
+
+#include <stddef.h>
+#include <stdint.h>
+#include <sys/types.h>
+
+/* Protocol families. */
+#define NSIM_PF_UNSPEC       0          /**< Unspecified protocol family.  */
+#define NSIM_PF_INET         1          /**< IP protocol family version 4. */
+
+/* Address families. */
+#define NSIM_AF_UNSPEC      NSIM_PF_UNSPEC   /**< Unspecified address family.   */
+#define NSIM_AF_INET        NSIM_PF_INET     /**< IP protocol family version 4. */
+
+/** Protocol numbers from IANA/BSD */
+enum nsim_net_ip_protocol {
+	NSIM_IPPROTO_IP = 0,            /**< IP protocol (pseudo-val for setsockopt() */
+	NSIM_IPPROTO_ICMP = 1,          /**< ICMP protocol   */
+	NSIM_IPPROTO_IGMP = 2,          /**< IGMP protocol   */
+	NSIM_IPPROTO_IPIP = 4,          /**< IPIP tunnels    */
+	NSIM_IPPROTO_TCP = 6,           /**< TCP protocol    */
+	NSIM_IPPROTO_UDP = 17,          /**< UDP protocol    */
+	NSIM_IPPROTO_IPV6 = 41,         /**< IPv6 protocol   */
+	NSIM_IPPROTO_ICMPV6 = 58,       /**< ICMPv6 protocol */
+	NSIM_IPPROTO_RAW = 255,         /**< RAW IP packets  */
+};
+
+/** Socket type */
+enum nsim_net_sock_type {
+	NSIM_SOCK_STREAM = 1,           /**< Stream socket type   */
+	NSIM_SOCK_DGRAM,                /**< Datagram socket type */
+	NSIM_SOCK_RAW                   /**< RAW socket type      */
+};
+
+#define NSIM_MSG_PEEK 0x02
+#define NSIM_MSG_TRUNC 0x20
+#define NSIM_MSG_DONTWAIT 0x40
+#define NSIM_MSG_WAITALL 0x100
+
+struct nsim_sockaddr {
+	uint16_t sa_family;      /* Address family */
+	char     sa_data[];      /* Socket address */
+};
+
+struct nsim_sockaddr_in {
+	uint16_t sin_family;     /* AF_INET */
+	uint16_t sin_port;       /* Port number */
+	uint32_t sin_addr;       /* IPv4 address */
+};
+
+struct nsim_sockaddr_in6 {
+	uint16_t sin6_family;    /* AF_INET6 */
+	uint16_t sin6_port;      /* Port number */
+	uint32_t sin6_flowinfo;  /* IPv6 flow info */
+	uint8_t  sin6_addr[16];
+	uint32_t sin6_scope_id;  /* Set of interfaces for a scope */
+};
```

Comment:
Why do we need to have separate nsim sockaddr structs here? I mean sockaddr is a sockaddr always, no?

Reply:
The problem that the introduction of `struct nsim_*` is trying to solve is to have a common layer between the two libc worlds:
- Native Simulator, which is compiled with host libc
- Zephyr, which is compiled with embedded libc

In order to pass `sockaddr` values between those two worlds, we need a structure that is defined the same way. This is where structs like `struct nsim_sockaddr_in6` are used, since they are locally defined and libc-independent. This is just to provide bridge between two conflicting libc implementations.